internal/git/gitsync_processor.go

```diff
@@ -0,0 +1,72 @@
+package git
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/storage/memory"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1 "github.com/numaproj-labs/numaplane/api/v1"
+)
+
+type Message struct {
+	Updates bool
+	Err     error
+}
+
+type GitSyncProcessor struct {
+	gitSync     v1.GitSync
+	channels    map[string]chan Message
+	k8Client    client.Client
+	clusterName string
+}
+
+func isCommitHash(hash string) bool {
+	match, _ := regexp.MatchString("^[0-9a-fA-F]{40}$", hash)
+	return match
+}
+
+func isValidTag(tag string) bool {
+	match, _ := regexp.MatchString(`^[vV]?\d+\.\d+\.\d+$`, tag)
+	return match
+}
+
+// reference can  can be a branch, a tag, or a commit hash
+func cloneRepository(repoUrl string, reference string) (*git.Repository, error) {
+	var referenceName plumbing.ReferenceName
+	if isValidTag(reference) {
+		referenceName = plumbing.ReferenceName(fmt.Sprintf("refs/tags/%s", reference))
+	} else if isCommitHash(reference) {
+		referenceName = plumbing.NewHashReference("", plumbing.NewHash(reference)).Name()
+	}
+	r, err := git.Clone(memory.NewStorage(), nil, &git.CloneOptions{
+		URL:           repoUrl,
+		SingleBranch:  true,
+		ReferenceName: referenceName,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error cloning the repository url %s", err)
+	}
+	return r, nil
+}
+
+func NewGitSyncProcessor(gitSync *v1.GitSync, k8client client.Client, clusterName string) (*GitSyncProcessor, error) {
+	channels := make(map[string]chan Message)
+	for _, repo := range gitSync.Spec.RepositoryPaths {
+		gitCh := make(chan Message)
```

Comment:
I'm thinking maybe we should add a length to the channel (maybe 5?) or otherwise we'll block if we try to send something and there's no space available. If we were just going to send a message to shutdown on here, it might be okay, but I think we will most likely also include a message to update path once we do Option 1 [here](https://docs.google.com/document/d/1d1xVkVwCT4sGijiL0ymboEcLuKTqYcg8mHh_9RjwECA/edit#heading=h.uaz74auunyxn).